### PR TITLE
feat(markup): added markup search

### PR DESCRIPTION
**Pull request by makhnatkin**

### CODE REVIEW

File: src/markup/codemirror/search-plugin/plugin.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    },
+    {},
+);
+
+export const removeDefaultSearch = search({
+    createPanel: () => ({
+        // Create an empty search panel
+        dom: document.createElement('div'),
+    }),
+});
```
Comment (d3m1d0v) [inline src/markup/codemirror/search-plugin/plugin.ts:135]: Lets provide this extension and search keymap with SearchPanelPlugin?

```js
    {
        provide: () => [
            keymap.of(searchKeymap),
            search({
                createPanel: () => ({
                    // Create an empty search panel
                    dom: document.createElement('div'),
                }),
            }),
        ],
    },
```

File: src/markup/codemirror/search-plugin/view/SearchPopup.tsx [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    const handleClose = () => {
+        setOpen(false);
+        onClose();
+    };
+
+    return (
+        <>
+            {anchor && (
+                <Portal container={anchor}>
+                    <SearchPopup open={open} onClose={handleClose} anchor={anchor} {...props} />
```
[the remark this answers is not in the document]
Comment (makhnatkin) [reply src/markup/codemirror/search-plugin/view/SearchPopup.tsx]: fixed

File: src/markup/codemirror/search-plugin/view/SearchPopup.tsx [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+interface PopupWithRefProps extends Omit<SearchPopupProps, 'open' | 'anchor'> {
+    anchor: HTMLElement | null;
+}
+export const PortalWithPopup: React.FC<PopupWithRefProps> = ({anchor, onClose, ...props}) => {
+    const [open, setOpen] = useState(true);
+
+    const handleClose = () => {
+        setOpen(false);
+        onClose();
+    };
```
[the remark this answers is not in the document]
Comment (makhnatkin) [reply src/markup/codemirror/search-plugin/view/SearchPopup.tsx]: fixed

File: src/markup/codemirror/search-plugin/plugin.ts (hunk 1 above)
Comment (makhnatkin) [reply src/markup/codemirror/search-plugin/plugin.ts]: fixed

File: src/markup/codemirror/search-plugin/plugin.ts [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    closeSearchPanel,
+    findNext,
+    findPrevious,
+    search,
+    searchKeymap,
+    searchPanelOpen,
+    setSearchQuery,
+} from '@codemirror/search';
+import {EditorView, PluginValue, ViewPlugin, ViewUpdate, keymap} from '@codemirror/view';
+import debounce from 'lodash/debounce';
```
Comment (d3m1d0v) [inline src/markup/codemirror/search-plugin/plugin.ts:14]: Lets import lodash utils from `src/lodash.ts`

File: src/markup/codemirror/search-plugin/view/SearchPopup.tsx [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    onSearchPrev?: (query: string) => void;
+    onSearchNext?: (query: string) => void;
+    onConfigChange?: (config: SearchConfig) => void;
+}
+
+const b = cn('search-card');
+
+const noop = () => {};
+
+const i18nSearch = i18n.bind(null);
```
Comment (d3m1d0v) [inline src/markup/codemirror/search-plugin/view/SearchPopup.tsx:40]: Why are you doing `bind()` for i18n?

File: src/markup/codemirror/search-plugin/plugin.ts [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+            this.handleSearchPrev = this.handleSearchPrev.bind(this);
+            this.handleSearchConfigChange = this.handleSearchConfigChange.bind(this);
+            this.setViewSearchWithDelay = debounce(this.setViewSearch, INPUT_DELAY);
+        }
+
+        update(update: ViewUpdate): void {
+            const isPanelOpen = searchPanelOpen(update.state);
+
+            if (isPanelOpen && !this.renderer) {
+                this.anchor = this.anchor ?? document.querySelector('.g-md-search-anchor');
```
Comment (d3m1d0v) [inline src/markup/codemirror/search-plugin/plugin.ts:63]: Lets provide query selector (`'.g-md-search-anchor'`) via options?

File: src/markup/codemirror/search-plugin/view/SearchPopup.tsx [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    onConfigChange: (config: SearchConfig) => void;
+    open: boolean;
+}
+
+export const SearchPopup: React.FC<SearchPopupProps> = ({open, anchor, onClose, ...props}) => {
+    const anchorRef = useRef<HTMLElement>(anchor);
+
+    return (
+        <>
+            {anchorRef && (
```
Comment (d3m1d0v) [inline src/markup/codemirror/search-plugin/view/SearchPopup.tsx:160]: `anchorRef` always exists

File: src/markup/codemirror/search-plugin/view/SearchPopup.tsx [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+SearchPopup.displayName = 'SearchPopup';
+
+interface PopupWithRefProps extends Omit<SearchPopupProps, 'anchor'> {
+    anchor: HTMLElement | null;
+}
+export const PortalWithPopup: React.FC<PopupWithRefProps> = ({anchor, open, onClose, ...props}) => {
+    return (
+        <>{anchor && <SearchPopup open={open} onClose={onClose} anchor={anchor} {...props} />}</>
+    );
+};
```
Comment (d3m1d0v) [inline src/markup/codemirror/search-plugin/view/SearchPopup.tsx:183]: You can replace this component with render function like [this](https://github.com/gravity-ui/markdown-editor/blob/9b808128a2042a91cfeead70a2d1113c4fe55961/src/markup/codemirror/files-upload-plugin/widget.tsx#L11C17-L11C29).
And remove import of React from `search-plugin/plugin.ts`

File: src/markup/codemirror/search-plugin/plugin.ts (hunk 4 above)
Comment (makhnatkin) [reply src/markup/codemirror/search-plugin/plugin.ts]: fixed

File: src/markup/codemirror/search-plugin/view/SearchPopup.tsx (hunk 5 above)
Comment (makhnatkin) [reply src/markup/codemirror/search-plugin/view/SearchPopup.tsx]: fixed

File: src/markup/codemirror/search-plugin/plugin.ts (hunk 6 above)
Comment (makhnatkin) [reply src/markup/codemirror/search-plugin/plugin.ts]: fixed

File: src/markup/codemirror/search-plugin/view/SearchPopup.tsx (hunk 8 above)
Comment (makhnatkin) [reply src/markup/codemirror/search-plugin/view/SearchPopup.tsx]: fixed

**Review by d3m1d0v — APPROVED** [review]

**Pull request merged**